pkg/scheduler/queue/queue.go

```diff
@@ -506,6 +535,41 @@ func (q *RequestQueue) processForgetDisconnectedQueriers() (resharded bool) {
 	return q.queueBroker.forgetDisconnectedQueriers(time.Now())
 }
 
+func (q *RequestQueue) SubmitRequestSent(req *SchedulerRequest) {
+	if req != nil {
+		select {
+		case q.requestsSent <- req:
+		case <-q.stopCompleted:
+		}
+	}
+}
+
+func (q *RequestQueue) processRequestSent(req *SchedulerRequest) {
+	if req != nil {
+		q.querierInflightRequests[req.Key()] = req
+		q.QueryComponentUtilization.IncrementForComponentName(req.ExpectedQueryComponentName())
+	}
+}
```

Comment:
Should this just be done as part of `enqueueRequestInternal`? Otherwise we could decide to send some requests to queriers, and then update the component utilisation based on those requests after all the decisions have been made. This means the decisions were made with lower utilisation than is true, which may lead us to overloading components.

(I'm not worried about potential delays decrementing utilisation - worst case scenario there is that we'll avoid a component we think is overloaded but actually isn't because we haven't observed the reduction in utilisation yet.)

Reply:
That is a possibility, but I have it this way because:
1. I prefer the symmetry and accuracy of two notification channels, with the outer process notifying sent only when it's actually sent, rather than sort of faking that it's sent inside the `RequestQueue`, then the outer process only notifies completion.
2. I do not believe we can reach a case where this difference matters much, as described below:

The case of "we make many dequeue decisions without returning to the `dispatcherLoop` to process the request completed notifications" can only happen when there were waiting queriers stacked up in `waitingQuerierConnsToDispatch`.
Waiting queriers stacked up can only occur meaning we went multiple `dispatcherLoop` cycles with waiting queriers and nothing in the queue for those queriers.

Firstly, during those loops to stack up connections into `waitingQuerierConnsToDispatch`, we will have processed the sent/completed notifications as well.

Secondly, remember that the scheduler-querier loops that have received a dequeued request cannot send it until the `RequestQueue` processes the sent notification and cannot loop to get another request until the `RequestQueue` processes the completed notification. This further restricts the cases where we can get waiting queriers stacked up.

Finally, a scenario with multiple of waiting queriers with nothing in the queue for them is unlikely to be a scenario where we will need to be concerned about query component utilization.

Reply:
note to self: if we do end up changing this we also need to a submitcompleted in this block, where we dequeue but found the request already canceled:

```golang
if schedulerReq.Ctx.Err() != nil {
			// remove from pending requests;
			// no need to SubmitRequestCompleted to RequestQueue as we had not yet SubmitRequestSent
			s.cancelRequestAndRemoveFromPending(schedulerReq.Key(), "request cancelled")
			lastUserIndex = lastUserIndex.ReuseLastTenant()
			continue
		}
```

Reply:
> The case of "we make many dequeue decisions without returning to the dispatcherLoop to process the request completed notifications" can only happen when there were waiting queriers stacked up in waitingQuerierConnsToDispatch.

This is not the scenario I had in mind.

Imagine there are many queries waiting in the queue, and three querier workers have just finished their queries (ie. we've already processed the `SubmitRequestCompleted` for the queries they've been working on). All other querier workers are busy.

Let's say we're also in a state where we would ideally only start one more store-gateway query, and all three workers would otherwise select a store-gateway query based on round-robin-ing between components.

At present, if I understand the code correctly, the following could happen:

* querier worker 1 calls `WaitForRequestForQuerier` and gets a store-gateway request, but the goroutine running `forwardRequestToQuerier` is then paused for whatever reason
* querier worker 2 calls `WaitForRequestForQuerier` and gets a store-gateway request, and its goroutine pauses
* querier worker 3 calls `WaitForRequestForQuerier` and gets a store-gateway request, and its goroutine pauses
* querier worker 1 only now calls `SubmitRequestSent`
* then querier worker 2 calls `SubmitRequestSent`
* and then querier worker 3 calls `SubmitRequestSent`

This is not what we want - we should only dispatch one store-gateway request.